### ISSUE TITLE
[COMPILER REQUEST]: NetBSD VAX: Update 10.5.0 based compiler, new 12.3.0 based compiler available

### ISSUE THREAD

**Issue by jbglaw** [compiler-explorer/compiler-explorer#5708]
### Compiler name

GCC

### Compiler version

`10.5.0 (NetBSD nb3 20231008)` and `12.3.0 (NetBSD nb1 20230729)`

### Compiler language

C/C++, others probably at some time as well

### Compiler homepage

https://gcc.gnu.org/ and https://www.netbsd.org/

### Motivation

Hi!

This refers to #4783: In the meantime, there were a good number of patches applied to "gcc.old" in NetBSD, which change quite a lot (for the better!) of details for the VAX backend. It would be cool to pick another current snapshot. (It would be sensible to add the snapshot date to the compiler name / dropdown. This can be found in `./external/gpl3/gcc.old/usr.bin/gcc/arch/vax/configargs.h` with something like this:
```
$ gcc_version="$(cat external/gpl3/gcc.old/dist/gcc/BASE-VER)"
$ nb_version="$(grep configuration_arguments external/gpl3/gcc.old/usr.bin/gcc/arch/vax/configargs.h |sed -e 's/^.*--with-pkgversion='"'"'\([^'"'"']*\)'"'"'.*$/\1/')"
$ nb_commit_timestamp="$(git show --no-patch --format='%ct')"
$ display_version="${gcc_version} (${nb_version} @${nb_commit_timestamp})"
```
)
...and then use `$display_version` to list this compiler in the drop-down?

Along with that new pile of patches for the "old" GCC used in NetBSD, the newer 12.x.y version (which contains a _hugh_ rework of the VAX backend) starts getting functional. It would be awesome to also have this compiler available. To _build_ it, it's enough to simply add `-V HAVE_GCC=12` to the `./build.sh` invocations. However, the above snippet to get GCC's version must be changed to look at the `gcc` directory (instead of `gcc.old`) for `${gcc_version}` and `${nb_version}`.

Thanks a lot,
  Jan-Benedict

**Comment by partouf** [comment]
Oh, nevermind. It seems I misunderstood the script

**Comment by jbglaw** [comment]
Ah, there's the script! Back in the day, I was suggesting changes/patches to @dkm. I didn't find the script ad hoc, otherwise I'd had sent a patch directly. Thanks for pointing out!

If you give me a short description on how that (these) scripts are called and how they're expected to place their output, I'd probably suggest a patch or prepare a pull request. The most prominent questions are:

* Is it called once manually, or every other day (e.g. by `cron` or similar)?
* ISTR that there was (at least) another file describing the compiler / tools. (How to call it, with which prefix, GCC-alike, ...) Where do I find that to also suggest a patch for the new compilers to be built?

Other notes:
* It would be a good idea to change the existing VAX compiler's name to contain the date of the snapshot it was taken. ISTR that the script basically creates (and stores) a tarball somewhere. The snapshot time taken from the tarball (or its contents) should be good enough.
* The current 10.5.0 compiler in NetBSD got a good number of patches, as well as the 12.3.0 based compiler. More work is being done, so it might be reasonable to fetch another snapshot in, say, half a year.

**Comment by dkm** [comment]
This script is called by hand within a docker image (Dockerfile in its parent dir). It's supposed to output its result in a tarball that will be stored in AWS. Usually, using a naming pattern that encodes everything we need (e.g. gcc-$target-$version). These are then referenced from the [infra repo conf](https://github.com/compiler-explorer/infra/blob/c9459657fb97ac481f5a46eb3a89be44f8a0e7ce/bin/yaml/cpp.yaml#L629). If we were to use arbitrary snapshot, we would probably need to pick something that connects the 2. I don't believe we have other compiler where we pick arbitrary snapshot and publish them in a similar manner as released compiler. But if that makes sense, that should be possible.

Sorry if I didn't implement changes you suggested during the initial introduction of the VAX compiler. I don't recall anything, but I probably forgot about them :(

**Comment by jbglaw** [comment]
@dkm That's valueable input, thanks a lot! So output filenames (to be stored in the S3 storage) goes along with the `name` list in the `cpp.yaml`? Need to check the pathes within the tarball.

As it seems, it should be easy to put all relevant information into the tarball name. Something indicating the snapshot (date? commit?) would probably be good to have. Can we possibly rename the original tarball? As NetBSD (upstream) uses CVS (which gets converted to `hg` and `git` every now and then), it would probably be most reasonable to not use a GIT commit ID, but a commit timestamp, to identify the snapshot. (As the GIT repo is used read-only, the conversion may depend on minor implementation details on the conversion process, thus commit IDs aren't really stable.

I'll hack that script (though I probably cannot completely test it?) to
* accept a `--new` option (chosing `gcc` instead of `gcc.old` and setting the required `GCC=12` variable), and
* allow for a `--timestamp` option to choose a specific commit.

That should result in tarballs named like `netbsd-vax-10.4.0-1699435461.tar.xz`, I think. (Have to have a closer look at the `${OUTPUT}` and `${S3OUTPUT}` handling though. Seems the variables I suggested back then were a bit crowded, or got crowded later on.)

That way, it should be easy re-creating the former GCC(.old) (that is already there) holding a proper timestamp (before all the recent patches) as well as the current (new+old) GCCs in the NetBSD tree.

PS: @dkm I think we both were happy at the time! So no regrets! It's actually a _good_ thing that there's been done so much work on GCC for VAX.

**Comment by jbglaw** [comment]
Here's my proposed updated script. I suggest to delete the current VAX compiler tarball (as it uses a different filename scheme from what the new script uses) and build two (three) compilers with it:

* `./build-vax-netbsd-2.sh --timestamp 1680300000` (build 10.4.0 based compiler with all its unpatched issues, matching the current compiler that's available on Compiler Explorer, based on an April 2023 timestamp)
* `./build-vax-netbsd-2.sh` (build current 10.5.0 based compiler with all the current new patches)
* `./build-vax-netbsd-2.sh --new` (build experimental 12.3.0 based compiler *once it also got the current patches*)

A recent build (`bash -x ./build-vax-netbsd-2.sh`, for current 10.5.0-based GCC) finished with:
```
+ export 'XZ_DEFAULTS=-T 0'
+ XZ_DEFAULTS='-T 0'
+ tar Jcf /tmp/nb-ce-build.2981360/gcc-vax--netbsdelf-10.5.0-1698193309.tar.xz -C /tmp/nb-ce-build.2981360 gcc-vax--netbsdelf-10.5.0-1698193309
+ '[' -n '' ']'
+ echo ce-build-output:/tmp/nb-ce-build.2981360/gcc-vax--netbsdelf-10.5.0-1698193309.tar.xz
ce-build-output:/tmp/nb-ce-build.2981360/gcc-vax--netbsdelf-10.5.0-1698193309.tar.xz
+ echo ce-build-status:OK
ce-build-status:OK
```
The resulting tarball contains (among others):
```
$ tar tJf /tmp/nb-ce-build.2981360/gcc-vax--netbsdelf-10.5.0-1698193309.tar.xz
[...]
gcc-vax--netbsdelf-10.5.0-1698193309/vax--netbsdelf-sysroot/lib/libc.so.12
gcc-vax--netbsdelf-10.5.0-1698193309/bin/vax--netbsdelf-gcc
gcc-vax--netbsdelf-10.5.0-1698193309/bin/vax--netbsdelf-objdump
[...]
```
Is that how it should look like? At least that'll allow to extract all of them next to each other without any conflicts. The `cpp.yaml` probably needs to be changed though.
I don't think that there's much interest in building other GCC versions from NetBSD sources for other architectures/machines, but should that arise, it'd be fairly easy to support. That script is obviously not tested with your Docker container, your S3 setup or the general Compiler Explorer environment. So please give it some careful checking before letting it run.

[build-vax-netbsd-2.sh.txt](https://github.com/compiler-explorer/compiler-explorer/files/13318845/build-vax-netbsd-2.sh.txt)

**Comment by dkm** [comment]
Thanks @jbglaw , that looks perfect. We never remove a compiler that has been made available, so I would preferably not change anything for the current 10.4.0. Knowing that your updated script can still build it is good though, thanks.

I can easily test and install a 10.5.0.

I'm not clear about the `once it also got the current patches`. Are you referring to what you mentioned before with `More work is being done, so it might be reasonable to fetch another snapshot in, say, half a year.`?

**Comment by jbglaw** [comment]
Since the first VAX compiler, a new (minor) version (10.4.0 -> 10.5.0) was imported, and a good pile of VAX-specific patches were applied. (Like ... in the last month or the like.) For the 10.5.x-based compiler, that work (applying patches) is (mostly?) finished, work is in progress to potentially upstream the work.
The next iteration is to apply the (already existing) patches to the 12.3.0 based compiler. That's real work in progress, I don't think NetBSD's current 12.3.x-based compiler will successfully build sysroot-required libs. (We'll have that in a few months I think.) That upcoming 12.3.x-based compiler contains (compared to 10.4.0 and 10.5.0) a hugh overhault of the whole VAX backend (from upstream), so we're eagerly waiting for it.
So bottom line: The first two suggested toolchains (when not removing the former VAX GCC, as you said, only that single one created by `./build-vax-netbsd-2.sh` without any further arguments) is something for "right now." The 12.3.0 compiler is subject to ... a new ticket I'll open in some months.

**Comment by dkm** [comment]
Thanks for the clarification! I'll take care of testing/building/configuring the 10.5.0 shortly :)

**Comment by dkm** [comment]
Am taking a quick look and will check your script within our docker image :+1:

**Comment by dkm** [comment]
```
./gcc-vax--netbsdelf-10.5.0-1699930998/bin/vax--netbsdelf-gcc --version
vax--netbsdelf-gcc (NetBSD nb3 20231008) 10.5.0
```
Works nicely! IIUC, I should name it with `10.5.0 (NetBSD nb3 timestamp)`, correct? I'll change the file name to use a readable date and that should be fine!

**Comment by jbglaw** [comment]
> ```
> ./gcc-vax--netbsdelf-10.5.0-1699930998/bin/vax--netbsdelf-gcc --version
> vax--netbsdelf-gcc (NetBSD nb3 20231008) 10.5.0
> ```
> 
> Works nicely! IIUC, I should name it with `10.5.0 (NetBSD nb3 timestamp)`, correct? I'll change the file name to use a readable date and that should be fine!

Correct, I used the epoch of the actual CVS "commit" (that is preserved in the CVS -> HG/GIT conversion.) That way, it's quite clear at which exact commit the snapshot was taken. The GCC version number (10.5.0) should change rarely (as it's the imported base version), but further patches for VAX might show up (though the 10.5.0 compiler is mostly settled by now, that's why I started this new PR.) But it is a sensible thing to do to (at least) have the snapshot's date, if not the complete (UTC?) timestamp.

**Comment by dkm** [comment]
Trying a compiler build:

https://github.com/compiler-explorer/infra/actions/runs/6876943020

**Comment by dkm** [comment]
Failed on the last s3 upload command, will fix that later today

**Comment by dkm** [comment]
Ok.. Using getopt was a nice idea, but it doesn't play well with our infra that makes some assumptions. I'll have to modify the script a bit more than expected.

**Comment by jbglaw** [comment]
Looking at the log:
```
2023-11-15T12:02:09.1390775Z ##[group]Run mkdir dist⏎
2023-11-15T12:02:09.1391126Z ^[[36;1mmkdir dist^[[0m⏎
2023-11-15T12:02:09.1391573Z ^[[36;1mchmod og+w dist # some builds don't run as root, so let anyone write here^[[0m⏎
2023-11-15T12:02:09.1392059Z ^[[36;1mecho ::group::Docker build^[[0m⏎
2023-11-15T12:02:09.1392534Z ^[[36;1mdocker run -v$(pwd)/dist:/dist --rm compilerexplorer/gcc-cross-builder \^[[0m⏎
2023-11-15T12:02:09.1393060Z ^[[36;1m  bash build-vax-netbsd.sh --s3 \^[[0m⏎
2023-11-15T12:02:09.1393403Z ^[[36;1m  /dist | tee output.log^[[0m⏎
2023-11-15T12:02:09.1393692Z ^[[36;1mecho ::endgroup::^[[0m⏎
2023-11-15T12:02:09.1394045Z ^[[36;1mecho ::group::Translate docker output into GH variables^[[0m⏎
2023-11-15T12:02:09.1394681Z ^[[36;1mgawk -e 'match($0, /^ce-build-(\w+):(.*)/, m) { print m[1] "=" m[2]; }' output.log >> "${GITHUB_OUTPUT}"^[[0m⏎
2023-11-15T12:02:09.1395197Z ^[[36;1mecho ::endgroup::^[[0m⏎
2023-11-15T12:02:09.1419530Z shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}⏎
```
...you just gave it `--s3 /dist`, while (I think---never ever worked with a s3 thingie) this should have been something like `--s3 s3://foo/dist` or the like.

**Comment by jbglaw** [comment]
https://github.com/compiler-explorer/compiler-explorer/issues/5708#issuecomment-1812748588

**Comment by dkm** [comment]
That's what I thought, but no (I may be wrong, @mattgodbolt or @partouf  can correct me). The bespoke workflow gets the path from the "echo" done by the script, /takes the basename/ and looks in the host filesytem (not within docker), see https://github.com/compiler-explorer/infra/blob/0cc82e34c76a6b1253723ee6e03dbf8ab0fea217/.github/workflows/bespoke-build.yaml#L63 . So having a nested output doesn't work as the nested part gets removed by the basename. I've tried some things yesterday but only realized what the correct fix was when shutting down the computer (and checking why my latest local run failed again). I'll fix that later tonight :)

**Comment by jbglaw** [comment]
On Thu, 2023-11-16 00:31:32 -0800, Marc Poulhiès ***@***.***> wrote:
> That's what I thought, but no (I may be wrong, @mattgodbolt or @partouf can
> correct me). The bespoke workflow gets the path from the "echo" done by the
> script, /takes the basename/ and looks in the host filesytem (not within
> docker), see https://github.com/compiler-explorer/infra/blob/0cc82e34c76a6b1253723ee6e03dbf8ab0fea217/.github/workflows/bespoke-build.yaml#L63 .
> So having a nested output doesn't work as the nested part gets removed by
> the basename. I've tried some things yesterday but only realized what the
> correct fix was when shutting down the computer (and checking why my latest
> local run failed again). I'll fix that later tonight :)

Ah, that's the point. My impression was that uploading to S3 would be
done inside the script, so there's no point at all in having a `--s3`
option to supply some kind of base URL.

  The actual fix would be to have an output *directory* where the
tarball is stored and point that to /dist (which probably is the
desired Docker mount to store it.)

MfG, JBG

--

**Comment by dkm** [comment]
Yes, I've changed the script to accept both an archive and a directory (as our other scripts are doing). But I messed up the `[ ] ||` ... So it's an easy fix that I'll test/merge later tonight :)
Thanks for your patience :)
